### ISSUE TITLE
Reviewer feedback: Move more RASP model design and construction details into main paper instead of leaving so much in the appendix

### ISSUE THREAD

**Issue by willy-b** [willy-b/RASP-for-ReCOGS#69]

**Comment by willy-b** [comment]
Also consider to emphasize other manually constructed models in the literature that are interesting.

Consider for example “Representing Rule-based Chatbots with Transformers” (Friedman et al 2024) which has as one of its objectives “constructing a Transformer that implements the ELIZA program, a classic, rule-based chatbot”, because “ELIZA illustrates some of the distinctive challenges of the conversational setting”. Seeing that paper helped me consider that implementing semantic parsing in RASP might be feasible, and thus actually demonstrating that semantic parsing can be implemented in RASP and providing our working code may be similarly useful to others for reference in applying RASP to their own objective.
(Note (Friedman et al 2024) and Friedman’s excellent other work (Friedman 2023) “Learning Transformer Programs” did not end up fitting in the final literature review for this work though I linked the latter from the repo attached to the paper I have wanted an excuse to add them back so may use this.)

Or for a much simpler setting, consider the Parity task, that of determining whether a sequence of 1s and 0s has an odd number of 1s or not (all references below are already cited and this is already mentioned in the literature review in the paper).
(Rumelhart et al. 1988) showed that a simple feedforward neural network could learn this by backpropagation on its own.
However, multiple authors, e.g. (David Chiang and Peter Cholak. 2022) and (Deletang et al., 2023) for example, have found that Transformers in practice cannot learn to solve the basic Parity task (though see (Zhou et al 2024) for a paper that uses RASP-L to show how the task can be modified so that Transformers can learn to solve it).
But in David Chiang and Peter Cholak 2022, "Overcoming a theoretical limitation of self-attention", the authors do artificially construct a handcrafted soft attention Transformer that can just barely solve Parity but only with confidence that is O(1/n) where n is the input length. This handcrafted solution is a useful reference if one wants to understand the limits of the attention operation used in Transformers to pass information between sequence positions.

Similarly we believe a handcrafted flat/non-hierarchical solution to the ReCOGS semantic parsing (and COGS as in the linked RASP-for-COGS implementation) may be useful to other authors as a reference approach (proof by construction that flat pattern matching with masking combined with decoder loop unrolling of recursion can get near-perfect scores on these benchmarks; despite the benchmark sentences coming from a PCFG with recursive grammar expansions).

**Comment by willy-b** [comment]
Last detail besides Table 1 added earlier was added in https://github.com/willy-b/RASP-for-ReCOGS/pull/95 .
